autode/opt/optimisers/crfo.py

```diff
@@ -76,28 +105,98 @@ def _step(self) -> None:
-        n_negative = sum(lmda < 0 for lmda in d2L_eigvals)
-        if not n_negative == m:
-            raise RuntimeError(
-                f"Constrained optimisation failed, ∇^2L has {n_negative} "
-                f" negative eigenvalues after RFO diagonal shift - "
-                f"should have {m}"
+        # RFO step in active space
+        o = m - n_satisfied_constraints
+        hessian = hessian[:, idxs][idxs, :]
+        gradient = self._coords.g[idxs]
+
+        d2l_eigvals = np.linalg.eigvalsh(hessian)
+        n_negative = sum(lmda < 0 for lmda in d2l_eigvals)
+        if n_negative != o:
+            raise OptimiserStepError(
+                f"Failed to calculated constrained RFO step, ∇^2L has "
+                f"{n_negative} negative eigenvalues after RFO diagonal"
+                f" shift - should have {o}"
             )
 
-        # Set all non-active components of gradient to zero
-        gradient = self._coords.g.copy()
-        gradient[self._coords.inactive_indexes] = 0.0
+        delta_s_active = -np.matmul(np.linalg.inv(hessian), gradient)
 
-        # take a quasi-Newton step
-        delta_s = -np.matmul(np.linalg.inv(hessian), gradient)
+        # form step in full space
+        delta_s = np.zeros(shape=(n + m,))
+        delta_s[idxs] = delta_s_active
 
-        # Set all the non-active components of the step to zero
-        delta_s[self._coords.inactive_indexes] = 0.0
+        # scale back to trust radius only on non-constraint modes
+        delta_s_q = delta_s[:n]
+        if np.linalg.norm(delta_s_q) > self.alpha:
+            delta_s = delta_s * self.alpha / np.linalg.norm(delta_s_q)
 
-        self._take_step_within_trust_radius(delta_s)
+        self._take_step_within_max_move(delta_s)
         return None
 
+    def _take_step_within_max_move(self, delta_s):
+        """
+        Take the step by converting internal coordinates to Cartesian
+        coordinates, and scaling back if the maximum movement of an
+        atom exceeds max_move
+
+        Arguments:
+            delta_s: The step in internal coordinates
+        """
+        assert self._coords is not None
+
+        self._coords.allow_unconverged_back_transform = True
+        new_coords = self._coords + delta_s
+        cart_delta = new_coords.to("cart") - self._coords.to("cart")
+        cart_displ = np.linalg.norm(cart_delta.reshape((-1, 3)), axis=1)
+        max_displ = np.abs(cart_displ).max()
+        if max_displ > self._maxmove:
+            logger.info(
+                f"Calculated step too large: max. displacement = "
+                f"{max_displ:.3f} Å, scaling down"
+            )
+            # Note because the transformation is not linear this will not
+            # generate a step exactly max(∆x) ≡ α, but is empirically close
+            factor = self._maxmove / np.abs(cart_displ).max()
+            delta_s = factor * delta_s
+
+        self._coords.allow_unconverged_back_transform = False
```

Comment:
perhaps should be set back to its initial value rather than `False`?

Reply:
@t-young31 I thought that when you set the coordinates, it must not allow unconverged transform (regardless of what it was set to before) since you want the coordinates in optimiser to be good. (eventually, i want to write code that will detect bad primitives and replace them automatically before the convergence issues arise.)